File: 2.1.py
class Node:
  def __init__(self, data, next):
    self.data = data
    self.next = next

class LinkedList():
  set = set()

  def __init__(self):
    self.head = None

  def add(self, item):
    tmp = Node(item, self.head)    
    self.head = tmp  

  def print(self):
    cur = self.head
    while cur != None:
      print('{} '.format(cur.data), end='')
      cur = cur.next    

   # O(n) space, O(1) time
  def remove_dups(self):
    cur = self.head
    prev = None
    while cur != None:
      if cur.data in self.set:
        prev.next = cur.next        
      else:
        self.set.add(cur.data)
        prev = cur
      cur = cur.next

    self.existing_nodes = set()
    return self.head

  # O(1) space, O(n^2) time
  def remove_dups_1(self):
    cur = self.head
    runner = None
    prev = None
    while cur != None:
      runner = cur.next
      prev = cur
      while runner != None:
        if runner.data == cur.data:
          prev.next = runner.next
        else:
          prev = runner
        runner = runner.next
      cur = cur.next

    return self.head

  # reverse linked list
  def reverse(self):
    current = self.head
    previous = None

    while current:
      next = current.next
      current.next = previous
      previous = current
      current = next

    self.head = previous
      
lst = LinkedList()

lst.add(8)
lst.add(490)
lst.add(234)

lst.print()
# lst.remove_dups_1()
print()
lst.reverse()
lst.print()

File: java/ch_2/2.1.java
import java.util.*;

class ListNode {
	int val;
	ListNode next;
	ListNode(int x) {
		val = x;
	}
}

class Solution {
	// O(n) time, O(n) space
	public static ListNode removeDups(ListNode head) {
		HashSet<Integer> hs = new HashSet<Integer>();
		ListNode prev = null;
		ListNode cur = head;
		ListNode next = null;
		while (cur != null) {
			// System.out.println("visit " + cur.val);
			next = cur.next;
			if (hs.contains(cur.val)) {
				// System.out.println(cur.val + " is in hs!");
				// delete cur
				prev.next = next;
				cur = next;
			} else {
				// System.out.println("add " + cur.val + " to hs");
				hs.add(cur.val);
				prev = cur;
				cur = next;
			}
		}
		return head;
	}

	// O(n) time, O(1) space
	public static ListNode removeDupsNoBuffer(ListNode head) {
		ListNode runner = null;
		ListNode prev = null;
		ListNode cur = head;
		while (cur != null) {
			runner = cur;
			prev = cur;
			while (runner.next != null) {
				runner = runner.next;
				if (runner.val == cur.val) {
					// delete current runner node
					prev.next = runner.next;
					runner = prev;
				} else {
					prev = runner;
				}
			}
			cur = cur.next;
		}
		return head;
	}

	public static void print(ListNode head) {
		ListNode cur = head;
		while (cur != null) {
			if (cur.next == null) {
				System.out.print(cur.val + "\n");
			} else {
				System.out.print(cur.val + " -> ");
			}
			cur = cur.next;
		}
	}

	public static void main(String[] args) {
		ListNode head = new ListNode(16);
		ListNode n1 = new ListNode(16);
		ListNode n2 = new ListNode(16);
		ListNode n3 = new ListNode(5);
		ListNode n4 = new ListNode(8);
		ListNode n5 = new ListNode(16);

		head.next = n1;
		n1.next = n2;
		n2.next = n3;
		n3.next = n4;
		n4.next = n5;
		
		print(head);
		removeDupsNoBuffer(head);
		print(head);
	}
}	
